### ISSUE TITLE
Window attention implementation

### ISSUE THREAD

**Issue by jzhang38** [thu-nics/DiTFastAttn#10]
I notice that the window attention is implemented using a mask:

https://github.com/thu-nics/DiTFastAttn/blob/3bf5b1aaf9976962d547cc4dd48d618b53e14f29/experiments/cfg_similarity_experiments/alter_attention_processor.py#L78C9-L78C19

am I correct to say this will not really bring speed up during training/inference as you are not really saving computation?

**Comment by Probe100** [comment]
This code was only used for exploring locality and CFG similarity in DiTs. For window attention, we utilize a different implementation in DiTFastAttn, which brings speed up in inference. Please check the code modules/fast_attn_processor for implementation details: https://github.com/thu-nics/DiTFastAttn/blob/3bf5b1aaf9976962d547cc4dd48d618b53e14f29/modules/fast_attn_processor.py.

**Comment by jzhang38** [comment]
I am able to locate this line:

https://github.com/thu-nics/DiTFastAttn/blob/3bf5b1aaf9976962d547cc4dd48d618b53e14f29/modules/fast_attn_processor.py#L122

So DiTFastAttn uses 1D sliding window instead of 2D? 

I also noticed you tried natten, both 1D and 2D. What makes you eventually decide to use flash attention with 1D sliding window?

**Comment by Probe100** [comment]
Yes, current we use 1D sliding window instead of 2D. We've tried natten 1d and natten 2d and we currently stick with flash attention with 1D sliding window since it achieves lowest latency.

**Comment by hahnyuan** [comment]
We use sliding window attention for WA-RS, which is the same as the sliding window attention used in Longformer and the 1D neighborhood attention. Our method can also be generalized to 2D local attention. However, we observe that 2D local attention is not efficient enough. This is because the GPU memory access of the K and V tensors is not sequential, resulting in extra data gathering overhead. We evaluate the inference latency of adopting Natten 2D attention and our sliding window attention. We use a kernel size of 127 for the sliding window, and a kernel size of 5 for the Natten 2D attention. The latency results are shown in Table below. Based on these results, we chose to use 1D local attention. 

| Kernel                    | Attention Latency |
|---------------------------|-------------------|
| Window Attention          | 0.828s            |
| Natten 2D Attention       | 0.926s            |

It is good point that to explore ways to improve the efficiency of the 2D attention kernel in the future.

**Comment by jzhang38** [comment]
@hahnyuan @Probe100 Thanks! This is extremely helpful.
